src/L2Penalty_alg.jl

```diff
@@ -0,0 +1,431 @@
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.status`: current status of the algorithm. Should be `:unknown` unless the algorithm has attained a stopping criterion. Changing this to anything will stop the algorithm, but you should use `:user` to properly indicate the intention.
+  - `stats.elapsed_time`: elapsed time in seconds.
+You can also use the `sub_callback` keyword argument which has exactly the same structure and in sent to `R2`.
+"""
+function L2Penalty(
+  nlp::AbstractNLPModel{T, V};
+	sub_solver = R2Solver,
+  kwargs...) where{ T <: Real, V }
+	if !equality_constrained(nlp) 
+		error("L2Penalty: This algorithm only works for equality contrained problems.")
+	end
+	solver = L2PenaltySolver(nlp,sub_solver = sub_solver)
+	stats = GenericExecutionStats(nlp)
+	solve!(
+		solver,
+		nlp,
+		stats;
+		kwargs...
+	)
+	return stats
+end
+
+function SolverCore.solve!(
+  solver::L2PenaltySolver{T, V},
+	nlp::AbstractNLPModel{T, V},
+	stats::GenericExecutionStats{T, V};
+	callback = (args...) -> nothing,
+	sub_callback = (args...) -> nothing,
+	x::V = nlp.meta.x0,
+	atol::T = √eps(T),
+	rtol::T = √eps(T),
+	neg_tol = eps(T)^(1/4),
+	ktol::T = eps(T)^(1/4),
+	max_iter::Int = 10000,
+	sub_max_iter::Int = 10000,
+	max_time::T = T(30.0),
+	max_eval::Int = -1,
+	sub_max_eval::Int = -1,
+	max_decreas_iter = 10,
+	verbose = 0,
+	sub_verbose = 0,
+	τ::T = T(100),
+	β1::T = τ,
+	β2::T = T(0.1),
+	β3::T = 1/τ,
+	β4::T = eps(T),
+	kwargs...,
+  ) where {T, V}
+    
+	reset!(stats)
+
+	# Retrieve workspace
+	h = NormL2(1.0)
+	ψ = solver.ψ
+	sub_ψ = solver.sub_ψ
+	sub_ψ.h = NormL2(τ)
+	solver.sub_solver.ψ.h = NormL2(τ)
+	
+	x = solver.x .= x
+	s = solver.s
+	s0 = solver.s0
+	shift!(ψ, x)
+	fx = obj(nlp, x)
+	hx = h(ψ.b)
+
+	if verbose > 0
+    @info log_header(
+      [:iter, :sub_iter, :fx, :hx, :theta, :xi, :epsk, :tau, :normx],
+      [Int, Int, Float64, Float64, Float64, Float64, Float64, Float64, Float64],
+      hdr_override = Dict{Symbol,String}(   # TODO: Add this as constant dict elsewhere
+        :iter => "outer",
+				:sub_iter => "inner",
+        :fx => "f(x)",
+        :hx => "‖c(x)‖₂",
+				:theta => "√θ",
+        :xi => "√(ξ/ν)",
+				:epsk => "ϵₖ",
+        :tau => "τ",
+        :normx => "‖x‖"
+      ),
+      colsep = 1,
+    )
+  end
+
+	set_iter!(stats, 0)
+	rem_eval = max_eval
+	start_time = time()
+	set_time!(stats, 0.0)
+	set_objective!(stats,fx + hx)
+	set_solver_specific!(stats,:smooth_obj,fx)
+	set_solver_specific!(stats,:nonsmooth_obj, hx)
+
+	local θ::T 
+	prox!(s, ψ, s0, 1.0)
+	θ = hx - ψ(s)
+
+	sqrt_θ = θ ≥ 0 ? sqrt(θ) : sqrt(-θ)
+	θ < 0 && sqrt_θ ≥ neg_tol && error("L2Penalty: prox-gradient step should produce a decrease but θ = $(θ)")
+	
+  atol += rtol * sqrt_θ # make stopping test absolute and relative
+	ktol = max(ktol,atol) # Keep ϵ₀ ≥ ϵ
+	tol_init = ktol # store value of ϵ₀ 
+
+	done = false
```

Comment:
Should compute the least-squares multipliers here. If by chance, the solver were initialized with a stationary point, we should exit before the main loop.